bee-network/bee-autopeering/src/request.rs

```diff
@@ -0,0 +1,202 @@
+        self.inner.write().expect("error getting write access")
+    }
+}
+
+pub(crate) struct RequestManagerInner {
+    version: u32,
+    network_id: u32,
+    source_addr: SocketAddr,
+    open_requests: HashMap<RequestKey, RequestValue>,
+}
+
+impl RequestManagerInner {
+    pub(crate) fn new_verification_request(
+        &mut self,
+        peer_id: PeerId,
+        peer_addr: IpAddr,
+        response_tx: Option<ResponseTx>,
+    ) -> VerificationRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<VerificationRequest>(),
+        };
+
+        let verif_req = VerificationRequest::new(self.version, self.network_id, self.source_addr, peer_addr);
+        let timestamp = verif_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::VerificationRequest,
+            &verif_req.to_protobuf().expect("error encoding verification request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        verif_req
+    }
+
+    pub(crate) fn new_discovery_request(
+        &mut self,
+        peer_id: PeerId,
+        response_tx: Option<ResponseTx>,
+    ) -> DiscoveryRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<DiscoveryRequest>(),
+        };
+
+        let disc_req = DiscoveryRequest::new();
+        let timestamp = disc_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::DiscoveryRequest,
+            &disc_req.to_protobuf().expect("error encoding discovery request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        disc_req
+    }
+
+    pub(crate) fn new_peering_request(
+        &mut self,
+        peer_id: PeerId,
+        response_tx: Option<ResponseTx>,
+        local: &Local,
+    ) -> PeeringRequest {
+        let key = RequestKey {
+            peer_id,
+            request_id: TypeId::of::<PeeringRequest>(),
+        };
+
+        let peer_req = PeeringRequest::new(local.public_salt().expect("missing public salt"));
+
+        let timestamp = peer_req.timestamp();
+
+        let request_hash = msg_hash(
+            MessageType::PeeringRequest,
+            &peer_req.to_protobuf().expect("error encoding peering request"),
+        );
+
+        let value = RequestValue {
+            request_hash,
+            expiration_time: timestamp + REQUEST_EXPIRATION_SECS,
+            response_tx,
+        };
+
+        let _ = self.open_requests.insert(key, value);
+
+        peer_req
+    }
+
+    pub(crate) fn pull<R: Request + 'static>(&mut self, peer_id: &PeerId) -> Option<RequestValue> {
```

Comment:
I think pull is a weird name but it's maybe just me, why not just follow the api and call it remove ?

Reply:
I dunno. I guess I liked the imagination of it at the time of coding this. 

Renamed.